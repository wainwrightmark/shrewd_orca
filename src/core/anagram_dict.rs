use itertools::{Itertools, MultiProduct};
use smallvec::SmallVec;
use std::{
    collections::{BTreeMap, HashMap},
    future::Future,
    iter::{FlatMap, Once},
    ops::Bound,
    str::FromStr,
};

use crate::core::prelude::*;

pub struct AnagramDict<'a> {
    pub words: BTreeMap<AnagramKey, Vec<Term<'a>>>,
}

impl<'a> From<TermDict<'a>> for AnagramDict<'a> {
    fn from(term_dict: TermDict<'a>) -> Self {
        let terms = term_dict.terms;

        Self::from(terms.into_iter())
    }
}

impl<'a, T: Iterator<Item = Term<'a>>> From<T> for AnagramDict<'a> {
    fn from(iter: T) -> Self {
        let groups = iter
            .sorted()
            .dedup()
            .filter_map(|term| AnagramKey::from_str(term.text).ok().map(|key| (key, term)))
            .into_group_map();
        let words = BTreeMap::from_iter(groups);

        AnagramDict { words }
    }
}

impl<'a> AnagramDict<'a> {
    fn solve_for_word(
        &self,
        word: &str,
        settings: SolveSettings,
    ) -> impl '_ + Iterator<Item = Vec<Term>> {
        let key = AnagramKey::from_str(word).unwrap();
        self.solve(key, settings)
    }

    fn solve(
        &self,
        key: AnagramKey,
        settings: SolveSettings,
    ) -> impl '_ + Iterator<Item = Vec<Term>> {
        let iterator = AnagramIterator::create(self, key, settings);

        let solutions = iterator.flat_map(|solution| {
            solution
                .into_iter()
                .map(|k| self.words.get(&k).unwrap().clone())
                .multi_cartesian_product()
        });

        solutions
    }
}

#[derive(Clone, Copy, Debug)]
pub struct SolveSettings {
    min_word_length: u8,
    max_words: usize,
}

impl Default for SolveSettings {
    fn default() -> Self {
        Self {
            min_word_length: 3,
            max_words: 3,
        }
    }
}

impl SolveSettings {
    pub fn allow(&self, key: &AnagramKey) -> bool {
        key.len >= self.min_word_length
    }
}

pub struct AnagramIterator<'a, 'b>
//TODO const N
{
    dict: &'b AnagramDict<'a>,
    stack: SmallVec<[(AnagramKey, AnagramKey, SmallVec<[AnagramKey; 5]>); 5]>,
    settings: SolveSettings,
}

impl<'a, 'b> AnagramIterator<'a, 'b> {
    pub fn create(dict: &'b AnagramDict<'a>, key: AnagramKey, settings: SolveSettings) -> Self {
        let mut stack = SmallVec::<[(AnagramKey, AnagramKey, SmallVec<[AnagramKey; 5]>); 5]>::new();
        stack.push((key, AnagramKey::EMPTY, SmallVec::<[AnagramKey; 5]>::new()));

        Self {
            dict,
            stack,
            settings,
        }
    }
}

impl<'a, 'b> Iterator for AnagramIterator<'a, 'b> {
    type Item = SmallVec<[AnagramKey; 5]>;

    fn next(&mut self) -> Option<Self::Item> {
        while !self.stack.is_empty() {
            let current_len = self.stack.len();

            if current_len > self.settings.max_words {
                self.stack.pop();
            } else if current_len == self.settings.max_words {
                let (current_key, _, used) = self.stack.pop().unwrap();

                if self.settings.allow(&current_key) && self.dict.words.contains_key(&current_key) {
                    let mut new_used = used.clone();
                    new_used.push(current_key);
                    return Some(new_used);
                }
            } else 
            {
                let (current_key, previous, used) = self.stack.last_mut().unwrap();

                if previous >= current_key {
                    //todo check previous squared
                    self.stack.pop();
                    continue;
                }

                if let Some((remainder, next_key)) = self
                    .dict
                    .words
                    .range((Bound::Excluded(*previous), Bound::Included(*current_key)))
                    .filter(|(&next_key, _)| self.settings.allow(&next_key))
                    .filter_map(|(&next_key, _)| {
                        (*current_key - next_key).map(|remainder| (remainder, next_key))
                    })
                    .next()
                {
                    previous.inner = next_key.inner;

                    if remainder.is_empty() {
                        let mut new_used = used.clone();
                        new_used.push(next_key);
                        return Some(new_used);
                    } else if next_key > remainder {
                        // if the remainder is in the dictionary, we have already passed it
                    } else if self.settings.allow(&remainder)
                        && self.settings.max_words > current_len
                    {
                        let mut new_used = used.clone();
                        new_used.push(next_key);
                        self.stack.push((
                            remainder,
                            AnagramKey {
                                inner: next_key.inner - 1,
                                len: next_key.len,
                            },
                            new_used,
                        ))
                    }
                } else {
                    self.stack.pop();
                }
            }
        }

        return None;
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use itertools::Itertools;

    // Note this useful idiom: importing names from outer (for mod tests) scope.
    use super::AnagramDict;
    use super::AnagramKey;
    use crate::core::prelude::*;

    #[test]
    fn test_solve() {
        let words = include_str!("common_words.txt")
            .split_ascii_whitespace()
            .map(|text| Term {
                text,
                tags: Default::default(),
                is_single_word: true,
            });

        let dict = AnagramDict::from(words);
        let solutions = dict.solve_for_word("clint eastwood", Default::default());

        let solutions_string = solutions
            .into_iter()
            .sorted_by_key(|x| x.len())
            .map(|s| s.into_iter().map(|t| t.text).join(" "))
            .join("; ");

        assert!(solutions_string.contains("west old action;"));        
    }

    #[test]
    fn test_solve_with_term_dict() {
        let term_dict = TermDict::from_term_data().unwrap();

        let dict = AnagramDict::from(term_dict);

        let solutions = dict.solve_for_word("clint eastwood", Default::default());

        let solutions_string = solutions
            .into_iter()
            .sorted_by_key(|x| x.len())
            .take(10)
            .map(|s| s.into_iter().map(|t| t.text).join(" "))
            .join("; ");

        assert_eq!(solutions_string, "Tito downscale; lot wainscoted; Watt colonised; twat colonised; cwt desolation; stint lacewood; Eliot downcast; Low anecdotist; owl anecdotist; town dislocate")
    }

    #[test]
    fn test_solve_basic() {
        let words = "act ire cat".split_ascii_whitespace().map(|text| Term {
            text,
            tags: Default::default(),
            is_single_word: true,
        });

        let dict = AnagramDict::from(words);

        let solutions = dict.solve_for_word("i react", Default::default());

        let solutions_string = solutions
            .into_iter()
            .sorted_by_key(|x| x.len())
            .map(|s| s.into_iter().map(|t| t.text).join(" "))
            .join("; ");

        assert_eq!(solutions_string, "ire act; ire cat");
    }

    #[test]
    fn test_duplicate_word() {
        let words = "cha".split_ascii_whitespace().map(|text| Term {
            text,
            tags: Default::default(),
            is_single_word: true,
        });

        let dict = AnagramDict::from(words);

        let solutions = dict.solve_for_word("chacha", Default::default());

        let solutions_string = solutions
            .into_iter()
            .sorted_by_key(|x| x.len())
            .map(|s| s.into_iter().map(|t| t.text).join(" "))
            .join("; ");

        assert_eq!(solutions_string, "cha cha");
    }
}
