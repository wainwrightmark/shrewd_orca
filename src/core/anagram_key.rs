use std::{
    fmt::{Debug, Display, Write},
    str::FromStr,
};

use anyhow::anyhow;
use prime_bag::PrimeBag128;

use super::prelude::Character;

#[derive(Copy, Clone, PartialEq, Eq, Hash)]
pub struct AnagramKey {
    len: u8,
    inner: prime_bag::PrimeBag128<Character>,
}

impl AnagramKey {
    pub fn is_empty(&self) -> bool {
        self.inner.is_empty()
    }

    pub fn empty() -> Self {
        Self {
            len: 0,
            inner: Default::default(),
        }
    }

    pub fn is_length_at_least(&self, l: u8) -> bool {
        self.len >= l
    }

    pub fn try_add(self, rhs: Self) -> Option<Self> {
        let inner = self.inner.try_sum(&rhs.inner)?;
        let len = self.len + rhs.len;
        AnagramKey { inner, len }.into()
    }

    pub fn try_sub(self, rhs: Self) -> Option<Self> {
        let inner = self.inner.try_difference(&rhs.inner)?;
        let len = self.len - rhs.len;
        AnagramKey { inner, len }.into()
    }
}

impl Ord for AnagramKey {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.inner.cmp(&other.inner)
    }
}

impl PartialOrd for AnagramKey {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Debug for AnagramKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let display = format!("{self}");

        f.debug_struct("AnagramKey")
            .field("txt", &display)
            .field("len", &self.len)
            .field("inner", &self.inner)
            .finish()
    }
}

impl Display for AnagramKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if self.is_empty() {
            f.write_char('!')?;
        } else {
            for char in self.inner.into_iter() {
                f.write_char(char.as_char())?;
            }
        }

        std::fmt::Result::Ok(())
    }
}

impl FromStr for AnagramKey {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, anyhow::Error> {
        let mut len: u8 = 0;

        let s = s.to_ascii_lowercase();

        let chars = s
            .chars()
            .filter(|c| c.is_ascii_lowercase())
            .flat_map(Character::try_from)
            .inspect(|_| len += 1);

        let inner = PrimeBag128::try_from_iter(chars).ok_or(anyhow!("String is too long"))?;

        Ok(AnagramKey { inner, len })
    }
}

#[cfg(test)]
mod tests {
    use super::AnagramKey;
    use std::str::FromStr;

    #[test]
    fn test_anagram_keys() {
        let clint_eastwood = AnagramKey::from_str("clint eastwood").unwrap();
        let old_west_action = AnagramKey::from_str("old west action").unwrap();
        assert_eq!(clint_eastwood, old_west_action);
    }

    #[test]
    fn test_add() {
        let clint = AnagramKey::from_str("clint").unwrap();
        let eastwood = AnagramKey::from_str("eastwood").unwrap();

        let clint_eastwood = (clint.try_add(eastwood)).unwrap();

        let old_west_action = AnagramKey::from_str("old west action").unwrap();
        assert_eq!(clint_eastwood, old_west_action);
    }

    #[test]
    fn test_sub() {
        let old_west_action = AnagramKey::from_str("old west action").unwrap();

        let eastwood = AnagramKey::from_str("eastwood").unwrap();

        let clint = AnagramKey::from_str("clint").unwrap();

        let subbed = (old_west_action.try_sub(eastwood)).unwrap();

        assert_eq!(clint, subbed);
    }
}
