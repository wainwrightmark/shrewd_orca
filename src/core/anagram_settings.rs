use itertools::{Itertools, MultiProduct};
use smallvec::SmallVec;
use std::{
    collections::{BTreeMap, HashMap},
    future::Future,
    iter::{FlatMap, Once},
    ops::Bound,
    str::FromStr,
};

use crate::{core::prelude::*};

#[derive(Clone, Copy, Debug)]
pub struct AnagramSettings {
    pub min_word_length: u8,
    pub max_words: usize,
}

impl Default for AnagramSettings {
    fn default() -> Self {
        Self {
            min_word_length: 3,
            max_words: 3,
        }
    }
}

impl AnagramSettings {
    pub fn allow(&self, key: &AnagramKey) -> bool {
        key.len >= self.min_word_length
    }
}