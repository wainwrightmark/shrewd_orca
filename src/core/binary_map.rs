use itertools::*;
use smallvec::SmallVec;
use std::{
    collections::BTreeMap,
    ops::{Bound, RangeBounds},
    str::FromStr,
};

use super::{anagram_key::AnagramKey, prelude::Homograph};

pub struct BinaryMap<Key, Value, const SIZE: usize> {
    keys: Vec<Key>,
    values: Vec<smallvec::SmallVec<[Value; SIZE]>>,
}

impl<Key: Ord, Value, const SIZE: usize> BinaryMap<Key, Value, SIZE> {
    pub fn get(&self, key: &Key) -> Option<&smallvec::SmallVec<[Value; SIZE]>> {
        match self.keys.binary_search(key) {
            Ok(index) => self.values.get(index),
            Err(_) => None,
        }
    }

    pub fn range(
        &self,
        range: impl RangeBounds<Key>,
    ) -> impl DoubleEndedIterator<Item = (&Key, &smallvec::SmallVec<[Value; SIZE]>)>
    {
        let start_bound = match range.start_bound() {
            Bound::Included(inc) => match self.keys.binary_search(inc) {
                Ok(b) => Bound::Included(b),
                Err(b) => Bound::Included(b),
            },
            Bound::Excluded(exc) => match self.keys.binary_search(exc) {
                Ok(b) => Bound::Excluded(b),
                Err(b) => Bound::Included(b),
            },
            Bound::Unbounded => Bound::Unbounded,
        };

        let end_bound = match range.end_bound() {
            Bound::Included(b) => match self.keys.binary_search(b) {
                Ok(b) => Bound::Included(b),
                Err(b) => Bound::Excluded(b),
            },
            Bound::Excluded(b) => match self.keys.binary_search(b) {
                Ok(b) => Bound::Excluded(b),
                Err(b) => Bound::Excluded(b),
            },
            Bound::Unbounded => Bound::Unbounded,
        };

        let bounds = (start_bound, end_bound);
        let keys = &self.keys[bounds];
        let values = &self.values[bounds];

        keys.iter().zip(values.iter())
    }

    pub fn contains_key(&self, key: &Key) -> bool {
        self.keys.binary_search(key).is_ok()
    }
}

impl<T: Iterator<Item = Homograph>, const SIZE: usize> From<T>
    for BinaryMap<AnagramKey, Homograph, SIZE>
{
    fn from(iter: T) -> Self {
        let groups = iter
            .sorted()
            .dedup()
            .filter_map(|term| AnagramKey::from_str(&term.text).ok().map(|key| (key, term)))
            .into_group_map();
        let words =
            BTreeMap::from_iter(groups.into_iter().map(|(k, g)| (k, SmallVec::from_vec(g))));

        Self {
            keys: words.keys().cloned().collect_vec(),
            values: words.into_values().collect_vec(),
        }
    }
}

// impl<Key, Value> BinaryMap<Key, Value> {
//     pub fn new(items: impl Iterator<Item = Value>, get_key : Fn(Value)-> Key)-> Self{
//         let mut values : Vec<Value> = items.collect();
//         values.sort_by_cached_key(get_key);
//         let keys : Vec<Key> = values.iter().map(get_key).collect();

//         Self { keys, values }
//     }
// }
