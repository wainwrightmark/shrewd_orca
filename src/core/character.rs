use num_derive::FromPrimitive;
use num_traits::FromPrimitive;
use prime_bag::*;
use serde::{Deserialize, Serialize};
use strum::{EnumCount, EnumIs, EnumIter};

#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    Hash,
    PartialOrd,
    Ord,
    Serialize,
    Deserialize,
    EnumCount,
    EnumIter,
    EnumIs,
    FromPrimitive,
)]
#[repr(u8)]

pub enum Character {
    E = 0,
    T = 1,
    A = 2,
    I = 3,
    N = 4,
    O = 5,
    S = 6,
    H = 7,
    R = 8,
    D = 9,
    L = 10,
    U = 11,
    C = 12,
    M = 13,
    F = 14,
    W = 15,
    Y = 16,
    G = 17,
    P = 18,
    B = 19,
    V = 20,
    K = 21,
    Q = 22,
    J = 23,
    X = 24,
    Z = 25,
}

impl PrimeBagElement for Character {
    fn into_prime_index(&self) -> usize {
        *self as usize
    }

    fn from_prime_index(value: usize) -> Self {
        FromPrimitive::from_usize(value).expect("Could not cast usize to character")
    }
}

impl std::fmt::Display for Character {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_char())
    }
}

impl Character {
    pub fn as_char(&self) -> char {
        match self {
            Character::A => 'A',
            Character::B => 'B',
            Character::C => 'C',
            Character::D => 'D',
            Character::E => 'E',
            Character::F => 'F',
            Character::G => 'G',
            Character::H => 'H',
            Character::I => 'I',
            Character::J => 'J',
            Character::K => 'K',
            Character::L => 'L',
            Character::M => 'M',
            Character::N => 'N',
            Character::O => 'O',
            Character::P => 'P',
            Character::Q => 'Q',
            Character::R => 'R',
            Character::S => 'S',
            Character::T => 'T',
            Character::U => 'U',
            Character::V => 'V',
            Character::W => 'W',
            Character::X => 'X',
            Character::Y => 'Y',
            Character::Z => 'Z',
        }
    }
}

impl TryFrom<char> for Character {
    type Error = &'static str;

    fn try_from(value: char) -> Result<Self, Self::Error> {
        match value {
            'a' | 'A' => Ok(Character::A),
            'b' | 'B' => Ok(Character::B),
            'c' | 'C' => Ok(Character::C),
            'd' | 'D' => Ok(Character::D),
            'e' | 'E' => Ok(Character::E),
            'f' | 'F' => Ok(Character::F),
            'g' | 'G' => Ok(Character::G),
            'h' | 'H' => Ok(Character::H),
            'i' | 'I' => Ok(Character::I),
            'j' | 'J' => Ok(Character::J),
            'k' | 'K' => Ok(Character::K),
            'l' | 'L' => Ok(Character::L),
            'm' | 'M' => Ok(Character::M),
            'n' | 'N' => Ok(Character::N),
            'o' | 'O' => Ok(Character::O),
            'p' | 'P' => Ok(Character::P),
            'q' | 'Q' => Ok(Character::Q),
            'r' | 'R' => Ok(Character::R),
            's' | 'S' => Ok(Character::S),
            't' | 'T' => Ok(Character::T),
            'u' | 'U' => Ok(Character::U),
            'v' | 'V' => Ok(Character::V),
            'w' | 'W' => Ok(Character::W),
            'x' | 'X' => Ok(Character::X),
            'y' | 'Y' => Ok(Character::Y),
            'z' | 'Z' => Ok(Character::Z),
            _ => Err("Invalid character"),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Default, Hash, Eq)]
pub struct CharacterMap<T>([T; 27]);

impl<T> CharacterMap<T> {
    pub fn get(&self, c: Character) -> &T {
        &self.0[c.into_prime_index()]
    }

    pub fn get_mut(&mut self, c: Character) -> &mut T {
        &mut self.0[c.into_prime_index()]
    }

    pub fn set(&mut self, c: Character, value: T) {
        self.0[c.into_prime_index()] = value;
    }

    pub fn enumerate(&self) -> impl Iterator<Item = (Character, &T)> {
        self.0
            .iter()
            .enumerate()
            .map(|(i, t)| (Character::from_prime_index(i), t))
    }
}
