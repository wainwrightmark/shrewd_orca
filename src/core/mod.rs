mod anagram_dict;
mod anagram_iterator;
mod anagram_key;
mod anagram_settings;
mod equation;
mod expression;
mod fixed_length_expression;
mod homograph;
mod many_expression_type;
mod pattern;
mod question;
mod solution;
mod solvable;
mod term_dict;
mod word_context;
mod word_query;

pub mod prelude {
    pub use crate::core::anagram_dict::*;
    pub use crate::core::anagram_iterator::*;
    pub use crate::core::anagram_key::*;
    pub use crate::core::anagram_settings::*;
    pub use crate::core::equation::*;
    pub use crate::core::expression::*;
    pub use crate::core::fixed_length_expression::*;
    pub use crate::core::homograph::*;
    pub use crate::core::many_expression_type::*;
    pub use crate::core::pattern::*;
    pub use crate::core::question::*;
    pub use crate::core::solution::*;
    pub use crate::core::solvable::*;
    pub use crate::core::term_dict::*;
    pub use crate::core::word_context::*;
    pub use crate::core::word_query::*;
}
