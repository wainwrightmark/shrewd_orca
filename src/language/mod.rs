mod parser;
mod phrase_expressions;
pub mod prelude {
    pub use crate::language::parser::*;
    pub use crate::language::phrase_expressions::*;
}
