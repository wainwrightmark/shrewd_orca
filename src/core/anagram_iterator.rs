use smallvec::SmallVec;
use std::ops::Bound;

use crate::core::prelude::*;

pub struct AnagramIterator<'b, const N: usize> {
    dict: &'b AnagramDict,
    stack: SmallVec<[(AnagramKey, Bound<AnagramKey>); N]>,
    used_words: SmallVec<[AnagramKey; N]>,
    settings: AnagramSettings,
}

impl<'b, const N: usize> AnagramIterator<'b, N> {
    pub fn create(dict: &'b AnagramDict, key: AnagramKey, settings: AnagramSettings) -> Self {
        let mut stack = SmallVec::<[(AnagramKey, Bound<AnagramKey>); N]>::new();
        stack.push((key, Bound::Included(key)));

        Self {
            dict,
            stack,
            settings,
            used_words: Default::default(),
        }
    }
}

impl<'b, const N: usize> Iterator for AnagramIterator<'b, N> {
    type Item = SmallVec<[AnagramKey; N]>;

    fn next(&mut self) -> Option<Self::Item> {
        while !self.stack.is_empty() {
            let top = self.stack.last_mut().unwrap();

            if let Some((remainder, next_key)) = self
                .dict
                .words
                .range((Bound::Unbounded, top.1))
                .rev()
                .filter(|(&next_key, _)| {
                    self.settings.allow_key(&next_key)
                    //&& self.settings.allow_word(possible_homographs)
                })
                .filter_map(|(&next_key, _)| {
                    (top.0 - next_key).map(|remainder| (remainder, next_key))
                })
                .next()
            {
                top.1 = Bound::Excluded(next_key);

                if remainder.is_empty() {
                    let mut new_used = self.used_words.clone();
                    new_used.push(next_key);
                    return Some(new_used);
                } else if self.settings.allow_key(&remainder) {
                    if self.settings.max_words  == Some(self.used_words.len() + 2) {
                        if remainder <= next_key && self.dict.words.contains_key(&remainder) {
                            //if(self.settings.allow_word(l))
                            {
                                let mut new_used = self.used_words.clone();
                                new_used.push(next_key);
                                new_used.push(remainder);
                                return Some(new_used);
                            }
                        }
                    } else if self.settings.max_words.is_none() || self.settings.max_words > Some(self.used_words.len() + 2) {
                        self.used_words.push(next_key);
                        self.stack.push((remainder, Bound::Included(next_key)))
                    }
                }
            } else {
                self.stack.pop();
                self.used_words.pop();
            }
        }

        None
    }
}
