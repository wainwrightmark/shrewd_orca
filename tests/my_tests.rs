use itertools::Itertools;
use word_playground::core::prelude::*;
use word_playground::language::prelude::*;

 use ntest::test_case;

    
 
 #[test_case("clint eastwood =a !n !j")]
 #[test_case("clint eastwood =a * *")]
 #[test_case("name * =a anagram *")] 
 #[test_case("5")]
    #[test_case("6 7")]
    #[test_case("red")]
    #[test_case("c?t fl?p")]
    #[test_case("clint eastwood")]
    #[test_case("6..7")]
    #[test_case("b?d")]
    #[test_case("b*d")]
    #[test_case("!n")]
    #[test_case("!v")]
    #[test_case("!a")]
    #[test_case("!j")]
    #[test_case("cat =a *")]
    

    fn test(input: String    ) {
        let context = WordContext::from_data();

        let p = word_lang_parse(input).unwrap();

        let solutions = p.solve(&context, &Default::default());

        let solutions_string = solutions
            .into_iter()
            .map(|s| s.get_text())
            .join("; ");

        insta::assert_snapshot!(solutions_string);
    }
