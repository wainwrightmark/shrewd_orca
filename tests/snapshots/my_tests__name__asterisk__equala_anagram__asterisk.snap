---
source: tests/my_tests.rs
expression: solutions_string
---
name agromania : anagram anomie; name cladogram : anagram mcleod; name cartilaginous : anagram licentious; name laggard : anagram geld; name gardant : anagram dent; name gardant : anagram tend; name vagrant : anagram vent; name ravaging : anagram given; name marginal : anagram limen; name gradable : anagram bleed
